pkg/informer/informer.go

```diff
@@ -0,0 +1,211 @@
+	OnAdd(gvr schema.GroupVersionResource, obj interface{})
+	OnUpdate(gvr schema.GroupVersionResource, oldObj, newObj interface{})
+	OnDelete(gvr schema.GroupVersionResource, obj interface{})
+}
+
+type GVREventHandlerFuncs struct {
+	AddFunc    func(gvr schema.GroupVersionResource, obj interface{})
+	UpdateFunc func(gvr schema.GroupVersionResource, oldObj, newObj interface{})
+	DeleteFunc func(gvr schema.GroupVersionResource, obj interface{})
+}
+
+func (g GVREventHandlerFuncs) OnAdd(gvr schema.GroupVersionResource, obj interface{}) {
+	if g.AddFunc != nil {
+		g.AddFunc(gvr, obj)
+	}
+}
+func (g GVREventHandlerFuncs) OnUpdate(gvr schema.GroupVersionResource, oldObj, newObj interface{}) {
+	if g.UpdateFunc != nil {
+		g.UpdateFunc(gvr, oldObj, newObj)
+	}
+}
+func (g GVREventHandlerFuncs) OnDelete(gvr schema.GroupVersionResource, obj interface{}) {
+	if g.DeleteFunc != nil {
+		g.DeleteFunc(gvr, obj)
+	}
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) Start(ctx context.Context) {
+	// Immediately discover types and start informing.
+	// TODO: Feed any failure to discover types into /readyz, instead of
+	// panicking.
+	if err := d.discoverTypes(ctx); err != nil {
+		klog.Fatalf("Error discovering initial types: %v", err)
+	}
+
+	// Poll for new types in the background.
+	ticker := time.NewTicker(pollTypesEvery)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				if err := d.discoverTypes(ctx); err != nil {
+					klog.Errorf("Error discovering types: %v", err)
+				}
+			}
+		}
+	}()
+}
+
+func (d *DynamicDiscoverySharedInformerFactory) discoverTypes(ctx context.Context) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	latest := sets.NewString()
+	rs, err := d.disco.ServerPreferredResources()
+	if err != nil {
+		return err
+	}
+	for _, r := range rs {
+		gv, err := schema.ParseGroupVersion(r.GroupVersion)
+		if err != nil {
+			return err
+		}
+		for _, ai := range r.APIResources {
+			if strings.Contains(ai.Name, "/") {
+				// foo/status, pods/exec, namespace/finalize, etc.
+				continue
+			}
+			if !ai.Namespaced {
+				// Ignore cluster-scoped things.
+				continue
+			}
+			if !sets.NewString([]string(ai.Verbs)...).HasAll("list", "watch") {
+				klog.V(2).Infof("resource %s.%s %s is not list+watchable: %v", gv.Group, gv.Version, ai.Name, ai.Verbs)
+				continue
+			}
+
+			latest.Insert(strings.Join([]string{ai.Name, gv.Version, gv.Group}, "."))
+		}
+	}
+
+	// Set up informers for any new types that have been discovered.
+	// TODO: Stop informers for types we no longer have.
+	newGVRs := latest.Difference(d.gvrs)
+	var merr error
+	for _, gvrstr := range newGVRs.UnsortedList() {
+		gvr, _ := schema.ParseResourceArg(gvrstr)
+		if gvr == nil {
+			multierr.AppendInto(&merr, fmt.Errorf("parsing GVR string %q: %w", gvrstr, err))
+			continue
+		}
+
+		inf := d.dsif.ForResource(*gvr).Informer()
+		if !cache.WaitForNamedCacheSync(gvr.String(), ctx.Done(), inf.HasSynced) {
```

Comment:
FYI, this will block forever if the cache never syncs. Could be bad (?)

Reply:
This was repurposed from @stevekuznetsov 's suggestion to do these checks. I can add a timeout if you'd like, but I'd like to add them everywhere consistently.

Reply:
We do this in the controllers and use the server's context for it, and we *do* see the server block forever on startup :)

Reply:
Yeah, this is a bit of a special/different case because DynamicDiscoverySharedInformerFactory acts both as a shared informer factory and a "controller" that handles watch events from informers.

And now that I'm looking at this a second time, we have to move this down after `go inf.Run(ctx.Done())`. Otherwise, if we leave it here, `inf.HasSynced` will immediately return true, because that's the (perhaps unfortunate) behavior for an unstarted informer.

Which makes me think more about the relationship between an informer, HasSynced, and someone checking if an informer has synced. In the "traditional" informer/controller split, we have this sequence:

1. Create shared informer factory
2. Create controller
3. Controller adds event handlers for an informer
4. (goroutine) Start controller
    1. Wait for all informers to sync
    2. Start workers
5. (goroutine) Start all informers
6. Informer does initial list, iterating:
    1. Each item is added to informer's store
    2. Event handler is called for each item (usually to enqueue into a work queue)
9. Controller is still waiting for the full list to be processed
10. Once full list is processed, informer.HasSynced() returns true, controller can start workers
11. Workers pop keys one at a time from the work queue

Here, though, we don't have a work queue, and instead we're passing the watch events straight through to `handler`. And the call to `cache.WaitForNamedCacheSync` is done is this goroutine, while the add/update/delete handler funcs are all called async from a different goroutine. Thus, I don't think there is any easy way to wait for the informer sync here that prevents the handler funcs from being called.

I think we may need to consider adding a workqueue here too. Happy to chat on Meet if a higher bandwidth discussion medium is helpful!